Advanced replication scheduling (cron)

erinkcochran87 opened the pull request:
This PR adds documentation for advanced replication scheduling using cron expressions.

erinkcochran87 commented:
@timvisher - I think this is ready for your 👀 .

I typically recommend people do a side-by-side of the rendered content and the raw files when they review docs, in case they want to leave comments. It's easier for me to address things that way and then it also doesn't get lost in Slack.

I've provided direct links (below) of the content that needs your blessing.

- This is the [rendered preview](https://deploy-preview-102--stitchdocs.netlify.com/docs/replication/replication-scheduling/advanced-cron-scheduling), or what the docs would look like on the live site.

And below are the files in this PR that make up that doc ^:

- [All of the main content is in this file.](https://github.com/stitchdata/docs/pull/102/files#diff-976ca4382adfe1f1628a4f4199712283)
- [Error messages are in this file](https://github.com/stitchdata/docs/pull/102/files#diff-7ffdd96c0eef5f147b6e97709175ec3e) - Creates the [**Troubleshooting** section](https://deploy-preview-102--stitchdocs.netlify.com/docs/replication/replication-scheduling/advanced-cron-scheduling) 
- [Field and value explanations and examples are in this file](https://github.com/stitchdata/docs/pull/102/files#diff-c79a7697763497a92241334cf4171852) - Creates the [**Formatting cron expressions**](https://deploy-preview-102--stitchdocs.netlify.com/docs/replication/replication-scheduling/advanced-cron-scheduling#cron-expression-field-values), [**Specifying days in cron expressions**](https://deploy-preview-102--stitchdocs.netlify.com/docs/replication/replication-scheduling/advanced-cron-scheduling#specifying-days), and [**Example replication schedules**](https://deploy-preview-102--stitchdocs.netlify.com/docs/replication/replication-scheduling/advanced-cron-scheduling#examples) sections.

Please complete the review I'm about to request from you, and feel free to ping me with questions :) Thank you! Let's wrap this up!

timvisher commented on _replication/scheduling/advanced-cron-scheduling.md:
```diff
@@ -0,0 +1,334 @@
+
+Using Advanced Scheduling, you can:
+
+1. **Run reports on a schedule**. For example: You need reports for a 9:00AM meeting you have every day. You could create a replication schedule that runs a replication job every night at 12:00AM, which would ensure fresh data is available in the morning when you need it. Reports could be scheduled to run at 7:00AM, giving you time before your meeting to review the data and resolve replication is ...
```
This is a great use case for anchor scheduling but not for cron scheduling. cron lets you express schedules that anchor can't capture (like every 15 minutes from 7 `7,22,37,52 * * * *`). I don't know how best to express that in the docs though but a customer reviewing both features would probably saying to themselves "huh. I'm pretty sure anchor does this too."

erinkcochran87 replied:
True - what if this were to be changed to be like `You have a meeting every Wednesday` versus `every day`?

erinkcochran87 replied:
Done.

timvisher commented on _replication/scheduling/advanced-cron-scheduling.md:
````diff
@@ -0,0 +1,334 @@
+[minute] [hour] [day of month] [month] [day of week]
+```
+
+### Formatting cron expressions {#cron-expression-field-values}
````
I find this section to be kind of confusing. :(

For instance the sentence:

```
The fields in a cron expression can either contain an asterisk (<code>*</code>), a list of integers separated by commas (ex: <code>1,2,3</code>), or an inclusive range, denoted by two numbers separated by a hyphen (ex: <code>1-5</code>). Fields may also contain both a range and a list.
```

sounds to me to indicate that each field can either be a single integer, a list of single integers, or a single range of integers. Also a field can contain a list/range pair.

This is different than the posix standard reads where each field can be an `*` or an element or a list of elements where an element is a single integer or an inclusive range of single integers. That properly captures that if you have an `*` it's all you can have, if you have an element it can be a single integer or integer range, and if you have a list it can be a list of any mixture of single integers or integer ranges all together.

I may only feel this way because I've been so deep on the posix specification for so long. :)

erinkcochran87 replied:
Could be! I think I'll throw this one in Slack and see how others interpret it.

erinkcochran87 replied:
Actually, I reworked this a bit:

```
The fields in a cron expression can contain:

- An asterisk (`*`), or
- A single integer (ex: `5`), or
- A list of integers separated by commas (ex: `1,2,3`), or
- An an inclusive range, denoted by two numbers separated by a hyphen (ex: `1-5`), or
- A combination of lists and inclusive ranges (ex: `1,5,10-20`)

Refer to the [Example replication schedules section](link) for examples, or the **Special characters tab** for more info on lists and ranges.
```

I also added clarification to the Special characters section that explains asterisks:

> **Note**: If used, an asterisk can be the only value in the field. For example: `*,30` would result in an error.

timvisher replied:
Stop me at any point if I'm being too nit-picky:

How 'bout:

```
The fields in a cron expression can contain:

- An asterisk (`*`), or
- A single integer (ex: `5`), or
- An inclusive range, denoted by two numbers separated by a hyphen (ex: `1-5`), or
- A list of any number of single integers and inclusive ranges separated by commas (`,`) (ex: `1,5,10-20,25,55-59`)
```

timvisher commented on _replication/scheduling/advanced-cron-scheduling.md:
```diff
@@ -0,0 +1,334 @@
+
+### Specifying days in cron expressions {#specifying-days}
+
+Stitch's implementation of cron uses [POSIX](http://pubs.opengroup.org/onlinepubs/7908799/xcu/crontab.html){:target="new"}, which, unlike other cron libraries, uses a unique method to handle the specification of days. This section will focus on the behavior of the `day of month`, `month`, and `day of week` fields.
```
I think this section is slightly misleading. It implies that there are cron libraries that don't have this combinatoric behavior when in fact all cron implementation do. It's specifically [Quartz](http://www.quartz-scheduler.org/documentation/quartz-2.x/tutorials/crontrigger.html) which is only cron-like that doesn't have it.

Again, I'm not sure how much of that detail to expose to our customers.

erinkcochran87 replied:
This might just be as simple as taking out the `unlike other cron libraries` bit, so it'd read something like:

> Stitch's implementation of cron uses POSIX, which uses a unique method to handle the specification of days.

timvisher replied:
That's possible. Maybe this as well:

> Stitch's cron scheduling implementation is POSIX compliant. The POSIX standard has subtle rules about how the final 3 fields of the cron expression affect the schedule on which the integration will run. This section…

Specifically `POSIX compliant` is how most software talks about whether or not it follows the POSIX specification.

erinkcochran87 replied:
Updated to:

> Stitch's implementation of cron is [POSIX](http://pubs.opengroup.org/onlinepubs/7908799/xcu/crontab.html) compliant. The POSIX standard has subtle rules about how the `day of month`, `month`, and `day of week` fields affect the schedule on which the integration will run. This section will focus on the behavior of those fields.

timvisher replied:
🙌

timvisher commented on _data/ui/cron-scheduling.yml:
````diff
@@ -0,0 +1,212 @@
+          [minute] [hour] 1 1 1
+          ```
+
+      - translation: "Run every day in the first month (January) and the first day of every week (Monday)"
````
I'm not sure if the english here properly highlights the combinatorics.

I might literally say something like "Run every day in the first month (January) and the first day of every week (Monday) including in January" but that sounds pretty awkward. Without that though I think it sounds too much like the options are exclusive.

erinkcochran87 replied:
I can see your point here, but `every day in the first month (January)` does include the first day of every week in January.

timvisher replied:
👍

timvisher commented on _replication/scheduling/advanced-cron-scheduling.md:
```diff
@@ -0,0 +1,334 @@
+## Example replication schedules {#examples}
+
+{% capture resource-callout %}
+**Want to test an expression?** Use [{{ cron.resource-names.crontab }}]({{ cron.resource-urls.crontab }}){:target="new"} to test expressions and see the schedule that will be created.
```
I think it's important to note here that crontab.guru implements several extensions that are not strictly posix compliant. So just because crontab.guru shows you the execution schedule of the cron expression doesn't mean that it will work in stitch.

erinkcochran87 replied:
Good call - I'll add that in.

timvisher commented on _data/ui/cron-scheduling.yml:
````diff
@@ -0,0 +1,212 @@
+      5-10,30 * * 1 *
+      ```
+
+  - translation: "Run every first day of the month and every week on Wednesday"
````
I think it'd be worth calling out "Run every minute on every first day of the month and every Wednesday

erinkcochran87 replied:
:+1:

timvisher commented on _data/ui/cron-scheduling.yml:
````diff
@@ -0,0 +1,212 @@
+      * * 1 * 3
+      ```
+
+  - translation: "Run every 15th day of the month in June and every Thursday at 5:00PM"
````
Similarly "Run at 5:00PM on every 15th day of the month in June and every Thursday"

timvisher commented on _replication/scheduling/advanced-cron-scheduling.md:
```diff
@@ -0,0 +1,334 @@
+
+1. **Advanced scheduling isn't available for all integrations**. Some database integrations don't currently support advanced scheduling. We are working on converting these integrations into Singer-powered taps, at which point advanced scheduling will be available.
+
+2. **Days of the week in specific months can't be specified** due to the implementation of cron Stitch uses. POSIX (the name of the cron library) [has a unique method for specifying days](#specifying-days).
```
posix is not a library but a standard http://pubs.opengroup.org/onlinepubs/9699919799/utilities/crontab.html

erinkcochran87 replied:
Fine, fine ;)

timvisher commented on _replication/scheduling/advanced-cron-scheduling.md:
```diff
@@ -0,0 +1,334 @@
+
+---
+
+## Troubleshooting {#troubleshooting-cron-errors}
```
Much of this will change once the new cron library is put in place.

erinkcochran87 replied:
Yep, I saw you mentioned that in Slack - do you have a lead on what those errors will say yet?

timvisher replied:
I don't have much specific information but the error generated _should_ allow us to say something like `5 fields are required (you specified '4')` or `6)` and `The hour field must be an integer between 0 and 23 (you specified '24')` or `'24,1-10')` or `'Yah/tva,.pn#')`.

timvisher replied:
I'll say this too, we should have the ability to generate whatever kind of error messages you think would be best, and you/lauren/mark/harrison are probably much better at phrasing error messages. So I'd make these look however you want them to look and whatever engineer works on it will make it happen. The error messages before were more because we were hoping someone else would do the error detection. Now we've been forced to do all of it so we can make it be whatever we want.

erinkcochran87 commented:
@timvisher I addressed your comments in my latest commits - wanna take a look? I think the only thing that's outstanding here are the refactored error messages.

erinkcochran87 commented:
Closing - we're addressing this in #307.
